shiny/ui/_theme.py

```diff
@@ -0,0 +1,521 @@
+            )
+
+        args: SassCompileArgs = {
+            "output_style": "compressed",
+            "include_paths": self._include_paths,
+            **args,
+        }
+
+        self._css = sass.compile(string=self.to_sass(), **args)
+
+        return self._css
+
+    # Third party theme-providers, e.g. shinyswatch, can override the next three methods
+    # to customize the HTML Dependency object that is returned by the theme or to
+    # provide pre-compiled CSS files.
+    def _dep_name(self) -> str:
+        """A method returning the name of the HTML dependency."""
+        return f"shiny-theme-{self.name or self._preset}"
+
+    def _dep_css_name(self) -> str:
+        """A method returning the name of the CSS file used in the HTML dependency."""
+        return "bootstrap.min.css"
+
+    def _dep_css_precompiled_path(self) -> str | pathlib.Path:
+        """A method returning the path to the precompiled CSS file."""
+        return path_pkg_preset(self._preset, self._dep_css_name())
+
+    def _read_precompiled_css(self) -> str:
+        path = self._dep_css_precompiled_path()
+        with open(path, "r") as f:
+            return f.read()
+
+    def _dep_create(self, css_path: str | pathlib.Path) -> HTMLDependency:
+        css_path = pathlib.Path(css_path)
+        return HTMLDependency(
+            name=make_valid_path_str(self._dep_name()),
+            version=self._version,
+            source={"subdir": str(css_path.parent)},
+            stylesheet={
+                "href": css_path.name,
+                "data-shiny-theme": self.name or self._preset,  # type: ignore
+            },
+        )
+
+    def _html_dependency_precompiled(self) -> HTMLDependency:
+        return self._dep_create(css_path=self._dep_css_precompiled_path())
+
+    def _html_dependency(self) -> HTMLDependency:
+        """
+        Create an `HTMLDependency` object from the theme.
+
+        Returns
+        -------
+        :
+            An :class:`~htmltools.HTMLDependency` object representing the theme. In
+            most cases, you should not need to call this method directly. Instead, pass
+            the `Theme` object directly to the `theme` argument of any Shiny page
+            function.
+        """
+        if self._can_use_precompiled():
+            return self._html_dependency_precompiled()
+
+        css_name = self._dep_css_name()
+        css_path = os.path.join(self._get_css_tempdir(), css_name)
+
+        if not os.path.exists(css_path):
+            with open(css_path, "w") as css_file:
+                css_file.write(self.to_css())
+
+        return self._dep_create(css_path)
+
+    def tagify(self) -> None:
+        raise SyntaxError(
+            "The `Theme` class is not meant to be used as a standalone HTML tag. "
+            "Instead, pass the `Theme` object directly to the `theme` argument of any "
+            "Shiny page function.",
```

Comment:
It'd be nice if this message was more express-aware (i.e., pass theme to `ui.page_opts()`)

Reply:
Is there a way to detect this at runtime?

Reply:
I'm not sure this is 100% correct/good idea, but I came up with this for my chat PR:

```python
def _express_is_active() -> bool:
    from ..express._run import get_top_level_recall_context_manager

    try:
        get_top_level_recall_context_manager()
        return True
    except RuntimeError:
        return False
```

Reply:
Done in 105a8f8443e1c9089c550b399a02089dc44f2562

Reply:
Oops, I missed your comment and include Express by including both instructions in the error message.